README.txt
SMSLib Sudden Motion Sensor Access Library Copyright (C) 2010 Daniel
Griscom, Suitable Systems

SMSLib is being distributed under the terms of the University of Illinois/
NCSA Open Source License: for more details see LICENSE.txt


About SMSLib:

SMSLib is an Objective C library giving access to the Sudden Motion
Sensor on recent Macintosh laptops. It includes default calibrations for
all such laptops, and includes a facility for storing customized
calibrations when desired.

Although SMSLib is written using Objective C, it is mostly vanilla C,
with Objective C used for logging support. It is not object-oriented;
there's little use in creating multiple instances of it, since they'd
all connect to the same single hardware device, and would have the same
state.


Using SMSLib:

1 - Call smsStartup() to initialize access to the SMS

2 - Call smsLoadCalibration() to load any stored calibration values

3 - Call smsGetData() to fetch calibrated SMS data. This can be done as
many times as you like; it takes 1ms to 2.5ms to return (on a MacBook Pro).
The acceleration record is a floating point value per axis, with 1.0
representing one gravity of acceleration.

4 - Call smsShutdown() when you have finished with SMSLib.

See smslib.m and SMSTester.m for an example of this.


Calibration:

Although SMSLib includes default calibrations for all flavors of the
SMS, each individual sensor will have slightly different zero and gain
values. To handle this, SMSLib allows you to store calibration values in
the system preferences, under the tag com.suitable.SeisMacLib.
SeisMaCalibrate is one application which uses this facility. Once
calibration values are saved, any application using SMSLib will
automatically use them.


Hardware:

SMSLib works on any Sudden Motion Sensor-equipped laptop, including the
most recent Core 2 Duo-equipped machines.

Although SMSLib was designed to hide the difference between different
hardware implementations of the SMS, there are still some visible
differences. The major ones are between the G4-based laptops and the
Intel-based laptops.

G4-based laptops all use SMSs which resolve 50 or so counts per gravity.
Many of them have a failure mode where SMSLib stops being able to access
the SMS until the next reboot. This seems to happen on the PowerBook5,6,
PowerBook5,7, PowerBook6,7 and PowerBook6,8 models; the PowerBook5,8 and
PowerBook5,9 models seem to be reliable.

Each of the G4-based models has different polarities for their axis
sensors; SMSLib corrects for this.

Intel-based laptops all use SMSs which resolve 250 or so counts per
gravity, and all have the same axis zeros and polarities. SMSLib access
is reliable on all Intel-based laptops.


Operating systems:

SMSLib works on any version of OS X that supports the host machine. It
has been tested on OS X 10.3.9 through 10.4.8, and I have a report that
it functions under early versions of 10.5.


About smsutil

smsutil is a command-line tool that uses SMSLib to output a series of
acceleration values. Its command-line arguments select which axes are
displayed, how often the values are fetched, and how many values should
be displayd. For example:

./smsutil -i0.005 -s09 -c1000

will output 1000 lines of data, with each line having tab-separated X, Y
and Z values, sampling every 5ms, covering 5 seconds of real time.

For further information on the command line options, type

./smsutil -h


About smslibtest

smslibtest is a command-line tool that tests SMSLib. It outputs a
detailed log of SMSLib's functions.


Further information:

For more information about SMSLib, see
<http://www.suitable.com/tools/smslib.html>

To calibrate your Sudden Motion Sensor, use SeisMaCalibrate:
<http://www.suitable.com/tools/seismacalibrate.html>

Or, you may enjoy the seismographic program SeisMac:
<http://www.suitable.com/tools/seismac.html>

To contact the author:
Daniel Griscom
Suitable Systems
1 Centre Street, Suite 204
Wakefield, MA 01880
(781) 665-0053
griscom@suitable.com

CHANGELOG.txt
SMSLib ChangeLog

SMSLib 1.8
- Fixed some memory leaks
- Added a number of new models, including the MacBookPro5,3 ,
	which has swapped X and Y axes and thus couldn't be calibrated
	properly
- Made code 64-bit capable

SMSLib 1.7
- Fixed Leopard bug writing corrupt calibrations
- Fixed bug crashing when reading corrupt calibrations
- Added comment that same calibration is used for both 15" and 17"
	variants of MacBook Pro Core 2 Duo with LED backlight, although
	latter has different signs for X and Y axes

SMSLib 1.6
- Improved calibration for MacBook Pro Core 2 Duo 15" with LED backlight

SMSLib 1.5
- Set smsutil and smslibtest to compile for both PowerPC and Intel machines

SMSLib 1.4:
- Improved debugging data (now shows 10Hz sine waves, decaying over 3 seconds,
		with X, Y and Z axes staggered by 1 second)
- Improved calibration with MacBook Pro Core 2 Duo 17" (Y and Z axes reversed)

SMSLib 1.3:

- Released under University of Illinois/NCSA Open Source License
- Greatly improved default calibration
- Cleaned up API


smstest.m
/*
 * smstest.m
 * smslib test command line tool
 *
 * SMSLib Sudden Motion Sensor Access Library
 * Copyright (c) 2010 Suitable Systems
 * All rights reserved.
 * 
 * Developed by: Daniel Griscom
 *               Suitable Systems
 *               http://www.suitable.com
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the
 * "Software"), to deal with the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 * 
 * - Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimers.
 * 
 * - Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimers in the
 * documentation and/or other materials provided with the distribution.
 * 
 * - Neither the names of Suitable Systems nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this Software without specific prior written permission.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR
 * ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS WITH THE SOFTWARE.
 *
 * For more information about SMSLib, see
 *		<http://www.suitable.com/tools/smslib.html>
 * or contact
 *		Daniel Griscom
 *		Suitable Systems
 *		1 Centre Street, Suite 204
 *		Wakefield, MA 01880
 *		(781) 665-0053
 *
 */

#import <Foundation/Foundation.h>
#import "smslib.h"
#import "SMSTester.h"

int main (int argc, const char * argv[]) {
	NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];

	SMSTester *test = [[SMSTester alloc] init];
	[test test];
	
	[pool release];
	
    return 0;
}


SMSTester.h
/*
 * SMSTester.h
 * smslib test class
 *
 * SMSLib Sudden Motion Sensor Access Library
 * Copyright (c) 2010 Suitable Systems
 * All rights reserved.
 * 
 * Developed by: Daniel Griscom
 *               Suitable Systems
 *               http://www.suitable.com
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the
 * "Software"), to deal with the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 * 
 * - Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimers.
 * 
 * - Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimers in the
 * documentation and/or other materials provided with the distribution.
 * 
 * - Neither the names of Suitable Systems nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this Software without specific prior written permission.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR
 * ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS WITH THE SOFTWARE.
 *
 * For more information about SMSLib, see
 *		<http://www.suitable.com/tools/smslib.html>
 * or contact
 *		Daniel Griscom
 *		Suitable Systems
 *		1 Centre Street, Suite 204
 *		Wakefield, MA 01880
 *		(781) 665-0053
 *
 */

@interface SMSTester : NSObject {
	NSMutableString *log;
	sms_acceleration accel;
}

- (void)logMessage: (NSString *)theString;
- (int)test;

@end


SMSTester.m
/*
 * SMSTester.m
 * smslib test class
 *
 * SMSLib Sudden Motion Sensor Access Library
 * Copyright (c) 2010 Suitable Systems
 * All rights reserved.
 * 
 * Developed by: Daniel Griscom
 *               Suitable Systems
 *               http://www.suitable.com
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the
 * "Software"), to deal with the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 * 
 * - Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimers.
 * 
 * - Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimers in the
 * documentation and/or other materials provided with the distribution.
 * 
 * - Neither the names of Suitable Systems nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this Software without specific prior written permission.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR
 * ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS WITH THE SOFTWARE.
 *
 * For more information about SMSLib, see
 *		<http://www.suitable.com/tools/smslib.html>
 * or contact
 *		Daniel Griscom
 *		Suitable Systems
 *		1 Centre Street, Suite 204
 *		Wakefield, MA 01880
 *		(781) 665-0053
 *
 */

#include <CoreFoundation/CoreFoundation.h>
#include <IOKit/IOKitLib.h>
#import "smslib.h"
#import "SMSTester.h"

@implementation SMSTester

// Log callback function used while starting up accelerometer
- (void)logMessage: (NSString *)theString {
	[log appendString:theString];
}

- (int)test {
	int i, length, result;
	log = [[NSMutableString alloc] init];
	
	[log appendString:@"\n========== SMSTest Report ==========\n"];
			
	// Start up SMS access
	result = smsStartup(self, @selector(logMessage:));
	if (result != SMS_SUCCESS) {
		// Couldn't start calibration.
		[log appendString:@"\n========== end ==========\n"];
		printf([log cString]);
		NSLog(log);
		return result;
	}
	
	// Fetch and display raw data
	[log appendString:@"\nRaw data:"];
	length = smsGetBufferLength();
	char *buffer = malloc(length);
	smsGetBufferData(buffer);
	for (i = 0; i < length; i++) {
		if (i % 16 == 0) {
			[log appendString:@"\n"];
		}
		[log appendString:[NSString stringWithFormat:@"%02.2hhx ", buffer[i]]];
	}
	[log appendString:@"\n\n"];
	
	// Load calibration
	[log appendString:@"Loading any saved calibration: "];
	if (smsLoadCalibration()) {
		[log appendString:@"success.\n\n"];
	} else {
		[log appendString:@"no saved calibration.\n\n"];
	}
	
	// Fetch and display calibration
	[log appendString:smsGetCalibrationDescription()];
	
	// Fetch and display one sample of calibrated acceleration
	[log appendString:@"\nFetching calibrated data: "];
	result = smsGetData(&accel);
	if (result != SMS_SUCCESS) {
		[log appendString:@"failed.\n"];
		[log appendString:@"\n========== end ==========\n"];
		printf([log cString]);
		NSLog(log);
		return result;
	}
	[log appendString:@"success.\n"];
	[log appendString:[NSString stringWithFormat:@"    X axis:%f\n", accel.x]];
	[log appendString:[NSString stringWithFormat:@"    Y axis:%f\n", accel.y]];
	[log appendString:[NSString stringWithFormat:@"    Z axis:%f\n", accel.z]];
	
	[log appendString:@"\n========== end ==========\n"];
	printf([log cString]);
	return 0;
}

@end


smsutil.m
/*
 * smsutil.m
 * smslib command line tool
 *
 * SMSLib Sudden Motion Sensor Access Library
 * Copyright (c) 2010 Suitable Systems
 * All rights reserved.
 * 
 * Developed by: Daniel Griscom
 *               Suitable Systems
 *               http://www.suitable.com
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the
 * "Software"), to deal with the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 * 
 * - Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimers.
 * 
 * - Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimers in the
 * documentation and/or other materials provided with the distribution.
 * 
 * - Neither the names of Suitable Systems nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this Software without specific prior written permission.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR
 * ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS WITH THE SOFTWARE.
 *
 * For more information about SMSLib, see
 *		<http://www.suitable.com/tools/smslib.html>
 * or contact
 *		Daniel Griscom
 *		Suitable Systems
 *		1 Centre Street, Suite 204
 *		Wakefield, MA 01880
 *		(781) 665-0053
 *
 */

#import <Foundation/Foundation.h>
#import "smslib.h"
#include <unistd.h>
#include <sysexits.h>

#define SMSUTIL_VERSION "1.2"

#define MAX_AXES_LENGTH (20)

// Prototypes
void SignalHandler(int sigraised);
long getUTime(void);
long getMTime(void);

static int running = YES;

int main (int argc, char ** argv) {
	// Configuration variables
	int count = 0;			// Number of samples; 0 means "go forever"
	float period = 1.0;		// Seconds between samples
	char axes[MAX_AXES_LENGTH + 1] = "xyz";
	BOOL calibrated = YES;
	BOOL help = NO;
	BOOL debugging = NO;
	int separator = ' ';
	
	// Other variables
	BOOL error = NO;		// Was there an argument error?
	long uPeriod;			// Microseconds between samples
	char *name;				// Base name of application (for messages)
	sms_acceleration accel;
	int c, x, result;
	NSAutoreleasePool *pool;
	long startMTime, lastUTime, thisUTime, waitUTime;
	
	// Get base name of application (for error messages)
	name = strrchr(argv[0], '/');
	name = (name) ? name + 1 : argv[0];
	
	// Parse command line options
	opterr = 0;	
	while ((c = getopt (argc, argv, ":a:c:i:s:uhd")) != -1) {
		switch (c) {
			case 'c':
				// Number of samples. Integer argument.
				if (sscanf(optarg, "%d", &count) != 1) {
					fprintf(stderr, "%s: bad argument to -c\n", name);
					error = YES;
				}
				break;
				
			case 'a':
				// Axes to be output. 
				if (strlen(optarg) > MAX_AXES_LENGTH) {
					// Arg too long
					fprintf(stderr, "%s: argument to -a limited to %d characters\n",
							name, MAX_AXES_LENGTH);
					error = YES;
				} else if (strspn(optarg, "xyzt") != strlen(optarg)) {
					// Arg not all "x", "y", "z" or "t"
					fprintf(stderr, "%s: argument to -a must only contain 'x's, 'y's, 'z's and 't's\n", name);
					error = YES;
				} else {
					strcpy(axes, optarg);
				}
				break;
				
			case 'i':
				// Seconds between samples. Can be floating point.
				if (sscanf(optarg, "%f", &period) != 1) {
					fprintf(stderr, "%s: bad argument to -i\n", name);
					error = YES;
				} else if (period < 0) {
					fprintf(stderr, "%s: argument to -i must be non-negative\n", name);
					error = YES;
				}
				break;
				
			case 'u':
				// Uncalibrated
				calibrated = NO;
				break;
				
			case 'd':
				// Debugging
				debugging = YES;
				break;
			
			case 'h':
				// Help
				help = YES;
				break;
			
			case ':':
				// Missing argument
				fprintf(stderr, "%s: option -%c requires an argument\n",
						name, optopt);
				error = YES;
				break;
			
			case 's':
				// Separator character
				if (sscanf(optarg, "%d", &separator) != 1) {
					fprintf(stderr, "%s: bad argument to -s\n", name);
					error = YES;
				} else if (separator > 255 || separator < 0) {
					fprintf(stderr, "%s: argument to -i must be 0-255\n", name);
					error = YES;
				}
				break;
				
			case '?':
				if (isprint (optopt))
					fprintf(stderr, "%s: unknown option `-%c'.\n", 
								name, optopt);
				else
					fprintf(stderr,
							 "%s: unknown option character `\\x%x'.\n",
							 name, optopt);
				error = YES;
				break;
				
			default:
				break;
		}
		
		// Was there a problem, or was help requested?
		if (help || error) {
			if (!error) {
				fprintf(stderr,
					"%s: prints values from Sudden Motion Sensor\n"
					"    version %s, SMSLib version %s\n"
					"    copyright (c) 2010 Daniel Griscom, Suitable Systems\n",
					name, SMSUTIL_VERSION, SMSLIB_VERSION);
			}
			fprintf(stderr,
				"usage: %s -[hud] [-i <period>] [-c <count>] [-a <axes>] [-s <char>]\n"
				"options:\n"
				"    -h          prints this message\n"
				"    -u          uncalibrated output\n"
				"    -d          debugging: hardware ignored, data are decaying sine waves\n"
				"    -i <period> seconds between samples (default: 1.0,\n"
				"                0 => sample as fast as possible)\n"
				"    -c <count>  number of samples (default: 0 => run forever)\n"
				"    -a <axes>   axes to be output (default: xyz,\n"
				"                can be in any order, use t for time in ms)\n"
				"    -s <char>   ASCII value of separator char (default: 32)\n",
				name);
			if (error) {
				return EX_USAGE;
			} else {
				return 0;
			}
		}
	}	// End parsing command line options
	
	// Set up signal handler for clean termination
    if (signal(SIGINT, SignalHandler) == SIG_ERR) {
        fprintf(stderr, "%s: could not establish new signal handler (err = %d)\n", name, errno);
		return EX_SOFTWARE;
	}
	
	// Now start doing the sampling!
	pool = [[NSAutoreleasePool alloc] init];
	if (debugging) {
		result = smsDebugStartup(nil, nil);
	} else {
		result = smsStartup(nil, nil);
	}
	if (result != SMS_SUCCESS) {
		fprintf(stderr, "Error: couldn't start SMS\n");
		[pool release];
		return EX_UNAVAILABLE;
	}
	
	if (calibrated) {
		smsLoadCalibration();
	}
	
	uPeriod = round(period * 1000000);
	lastUTime = getUTime();
	startMTime = getMTime();
	
	// Sample loop
	while (running) {
		if (calibrated) {
			smsGetData(&accel);
		} else {
			smsGetUncalibratedData(&accel);
		}
		// Field loop
		for (x = 0; x < strlen(axes); x++) {
			c = axes[x];
			if (x > 0) {
				printf("%c", separator);
			}
			switch (c) {
				case 'x':
					if (calibrated) {
						printf("%f", accel.x);
					} else {
						printf("%d", lrint(accel.x));
					}
					break;
					
				case 'y':
					if (calibrated) {
						printf("%f", accel.y);
					} else {
						printf("%d", lrint(accel.y));
					}
					break;
				
				case 'z':
					if (calibrated) {
						printf("%f", accel.z);
					} else {
						printf("%d", lrint(accel.z));
					}
					break;
					
				case 't':
					// Show milliseconds since program started
					printf("%d", getMTime() - startMTime);
					break;
			}
		}	// End field loop
		
		printf("\n");
		if (count > 0 && --count == 0) {
			// Time to stop sampling
			break;
		}
		
		// Yes, there's overflow all over the place here, but it doesn't
		// matter.
		thisUTime = getUTime();
		waitUTime = lastUTime + uPeriod - thisUTime;
		if (waitUTime > 0) {
			usleep(lastUTime + uPeriod - thisUTime);
		}
		lastUTime += uPeriod;
		// Make sure we don't get too far behind
		if (lastUTime < thisUTime) {
			lastUTime = thisUTime;
		}
	}	// End sample loop
	
	[pool release];
	
	return 0;
} // End main()

// Returns time in microseconds, clipped to a long
long getUTime() {
	struct timeval t;
	gettimeofday(&t, 0);
	return (t.tv_sec * 1000000 + t.tv_usec);
}

// Returns time in milliseconds, clipped to a long
long getMTime() {
	struct timeval t;
	gettimeofday(&t, 0);
	return (t.tv_sec * 1000 + t.tv_usec / 1000);
}

// Called on SIGINT (control-C)
void SignalHandler(int sigraised)
{
	// Stop runloop
    running = NO;    
}
